[Infra UI] ECS Migration

simianhacker opened the pull request:
This PR migrates the fields outline in elastic/kibana#27284 and #26618. The Metadata endpoint had to be refactored because `event.dataset` format is  an array of strings instead concatenating `metricset.module` and `metricset.name`.

### Reviewers

- @skh I added you to this review because of the refactor to the Metadata endpoint
- @ruflin This PR should give you something to work against.

elasticmachine commented:
Pinging @elastic/infrastructure-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5922/)

ruflin commented on x-pack/plugins/infra/server/lib/domains/metadata_domain/metadata_domain.ts:
```diff
@@ -48,17 +48,9 @@ export class InfraMetadataDomain {
-      })
-      .reduce((a: string[], b: string[]) => a.concat(b), []);
+    const metadata = buckets.map(module => {
+      return module.key;
```
Not sure fully understand this change. Is it because of the dataset? If yes, should this be dataset instead of module.key?

simianhacker replied:
`module` is an aggregation bucket, the value we are looking for is stored on the `key` attribute.

skh replied:
It might make sense to rename the argument in the anonymous function passed to `map()` to `dataset` as this has indeed nothing to do with modules any more?

skh commented:
Do we have data to test this with in the dev cluster?

I used this graphql query and it returns no data:
```
{
  source(id: "default") {
    id
    metadataByNode(nodeId: "demo-stack-apache-01", nodeType: host) {
      id
      features {
        name
        source
      }
    }
  }
}
```

simianhacker commented:
@skh The `demo-stack` cluster is not running the latest versions of metricbeat/filebeat which have the correct versions. For now you will just need to run them locally.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6048/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6079/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6150/)

spalger commented:
retest

skh commented:
I'm having a plain ubuntu vagrant box shipping `system` data from a 7.0.0-alpha2 metricbeat package to the dev cluster, but this query still returns no features:

```
{
  source(id: "default") {
    id
    metadataByNode(nodeId: "skh-ubuntu-bionic64-01", nodeType: host) {
      id
      features {
        name
        source
      }
    }
  }
}
```

In addition, the waffle map does not show that host either -- it's completely empty.

Can you describe in more detail how this should be tested?

simianhacker commented:
Looks like I'm going to need to re-archive the data for this test, since the current data is in the 6.x format.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6420/)

simianhacker commented:
@ruflin I'm using the latest nightly snapshot of Metricbeat and `host.hostname` doesn't exist.

ruflin commented:
@simianhacker The next snapshot will contain it as I just merged https://github.com/elastic/beats/pull/9943 But I recommend to use `host.name` for the above anyways as it's always around.

ruflin commented:
@skh For testing:

* Use a 6.6 snapshot of Metricbeat / Filebeat
* Test it against this branch and check if all fields show up as expected. We should test this in all 3 environments: Host, Docker, K8s

What we are looking for is if any fields / data is missing.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6515/)

skh commented:
Ok, so the metadata endpoint works for data sent with metricbeat-6.6.0 BC2, but not with 7.0.0-alpha2.

Also, as mentioned above, the host running 7.0.0 alpha2 doesn't even appear on the waffle map.

Is this the intended behaviour?

ruflin commented:
It should work with both versions. Quite a few changes happened since alpha 2. Can you try a snapshot build from here (linux): https://beats-ci.elastic.co/job/elastic+beats+master+multijob-package-linux/gcsObjects/ (os x): https://beats-ci.elastic.co/job/elastic+beats+master+multijob-package-darwin/gcsObjects/

skh commented:
@ruflin Thanks for the pointer! Data from today's `7.0.0` snapshot is displayed correctly as well.

That leaves the question if it is intended behaviour that the waffle map now only shows nodes with the new ECS fields present. For a user with older data it will look as if their whole infrastructure disappeared until they update the beats.

simianhacker commented:
I'm removing the review tag off this PR it's not quite ready. The test data needs to be updated with ECS compatible data.

@skh For testing I've been working on getting the vagrant stuff updated. Unfortunately, the URL where we were pulling down the snapshot packages (via Ansible) are not working with the `7.0.0-SNAPSHOT-amd64` variants. That took me a while to get working right yesterday and I finally just removed the download portion, the admin will have to do that manually.

@ruflin Just to be clear we need to run this against `6.6.0` and `7.0.0` data for testing?

ruflin commented:
@simianhacker Yes. Most important one at the moment is 6.6.0 but it needs to work with both.

ruflin commented:
@skh Infra UI 7.0 will only be backward compatible with data coming from 6.6 and 6.7. For older data we will provide a migration assistant that can apply the needed alias (at least that is the plan so far).

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6860/)

ruflin commented:
I did a bit more testing on this with Elastic-Stack master (this PR) against 6.6 Metricbeat on Docker. The problem here is the same as it was before for the `host.hostname`. When hovering it shows `null` instead of the docker name as it probably tries to pull the data from `_source` but as it is an alias it can't access it. I assume the same problem will apply to K8s but will also test there to confirm.

With the change to `host.name`, the host overview shows the correct data.


![screenshot 2019-01-14 at 14 32 37](https://user-images.githubusercontent.com/244900/51115836-a136b480-1809-11e9-98a1-679a09e1e107.png)

ruflin commented on x-pack/plugins/infra/server/lib/adapters/nodes/constants.ts:
```diff
@@ -11,5 +11,5 @@ export const NODE_REQUEST_PARTITION_FACTOR = 1.2;
   [InfraNodeType.host]: 'host.name',
   [InfraNodeType.pod]: 'kubernetes.pod.name',
-  [InfraNodeType.container]: 'docker.container.name',
+  [InfraNodeType.container]: 'container.name',
```
I see that `kubernetes.pod.name` did not change, so the source issue probably only exists for the container.name.

simianhacker commented:
@ruflin Where are you getting the Docker data to test?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7138/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7140/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7294/)

simianhacker commented:
@ruflin @skh I had to re-arrange the test data due to some changes upstream with Elasticsearch. There are now sub folders for the `infra` fixtures. The `legacy` folder contains the old test data we used originally, but with an updated compatible `mapping.json`. There are now two other folders, `6.6.0` and `7.0.0`, each with data from the respective versions of Metricbeat and Filebeat (with their mappings fixed, see below). I added tests to both the waffle map and metadata endpoints to make sure they continue to behave according to the goals of this PR. I'm sure there will be more work to do when this PR is back ported since I believe the 6.x version all run against 6.x versions of Elasticsearch and I doubt the data is compatible between them, the compatibility issues stems from Elasticsearch dropping support of `types`.

As far as getting test data.... there isn't an easy answer to this problem right now because you basically have to use the `auto-discovery` setup that @ruflin sent me to test the 6.6.0 data but you have to put that data into `7.0.0-SNAPSHOT` version of Elasticsearch, which can be very challenging. I had to export it using the `scripts/es_archiver.js` from `6.6.0`, then edit the `mappings.json` file and `data.json.gz` and reload it into a `7.0.0-SNAPSHOT` version of Elasticsearch using the `es_archiver` script.

Ideally I would like to have a `6.6.0` and `7.0.0` test for hosts, containers and pods but the level of effort to get that data exported, fixed and working is pretty high. I feel like what I have thus far gives us the minimum level of confidence in this PR.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7323/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7337/)

simianhacker commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7360/)

simianhacker commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7375/)

simianhacker commented:
retest

ruflin commented:
I tested this again with metricbeat 6.7 SNAPSHOT and this PR + master ES. Docker and host seem to work es expected. As Metricbeat 6.6 can't ship to 7 at some point we need to test this by migrating an ES instance with data.

Will try to test also with k8s setup again.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8771/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9350/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9488/)

ruflin commented on x-pack/plugins/infra/server/lib/domains/log_entries_domain/builtin_rules/filebeat_apache2.test.ts:
```diff
@@ -12,12 +12,12 @@ describe('Filebeat Rules', () => {
   test('Apache2 Access', () => {
     const event = {
       'apache2.access': true,
```
The apache2 module got rename to apache so did the fields that are not in ECS.

ruflin commented on x-pack/plugins/infra/server/lib/domains/log_entries_domain/builtin_rules/filebeat_apache2.test.ts:
```diff
@@ -12,12 +12,12 @@ describe('Filebeat Rules', () => {
+      'url.original': '/faqs',
+      'http.version': '1.1',
+      'http.response.status_code': '200',
       'apache2.access.body_sent.bytes': 1024,
```
This field changed to `http.request.body.bytes`

weltenwort replied:
I fixed those fields in a PR to this branch.

ruflin commented on x-pack/plugins/infra/server/lib/domains/log_entries_domain/builtin_rules/filebeat_nginx.test.ts:
```diff
@@ -12,61 +12,61 @@ describe('Filebeat Rules', () => {
+      'http.request.method': 'GET',
+      'url.original': '/faq',
+      'http.version': '1.1',
       'nginx.access.body_sent.bytes': 1024,
```
http.request.body.bytes

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9801/)

weltenwort commented on x-pack/test/api_integration/apis/infra/constants.ts:
```diff
@@ -0,0 +1,20 @@
+      max: 1547579090048,
+    },
+  },
+};
```
that's a great idea... can't remember how often I had to look those up

weltenwort commented on x-pack/test/api_integration/apis/infra/index.js:
```diff
@@ -5,6 +5,7 @@
  */
 
 
+console.log("Did you load?");
```
I guess this is a left-over `console.log`?

simianhacker replied:
oopsie!

simianhacker commented:
@weltenwort I agree it's a hard one to judge. I think we should merge it as well then fix any bugs we find. For the general use cases, everything I've tested so far works.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10432/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10442/)

ruflin commented:
++ on getting merging this.
